### PR TITLE
Bring back List filter/filterNot optimization to 2.13

### PR DESCRIPTION
**Pull request by cryptom-dev**
This change was brought to 2.12 from https://github.com/scala/scala/pull/5664/files
but as it broke MIMA check, it was reverted from: https://github.com/scala/scala/commit/76babbb6728e5daf4fb95444273f6f8e0a4098d7

### CODE REVIEW

**Comment by retronym** [comment]
The test failure:

```
[test] !!  615 - run/repl-colon-type.scala                 [output differs]
[test] % diff /home/jenkins/workspace/scala-2.13.x-validate-main@2/test/files/run/repl-colon-type-run.log /home/jenkins/workspace/scala-2.13.x-validate-main@2/test/files/run/repl-colon-type.check
[test] @@ -79,7 +79,7 @@ TypeRef(
[test]      )
[test]      TypeRef(
[test]        TypeSymbol(
[test] -        sealed abstract class List[+A] extends AbstractSeq[A] with LinearSeq[A] with Product with GenericTraversableTemplate[A,List] with LinearSeqOptimized[A,List[A]] with FilteredTraversableInternal[A,List[A]] with Serializable
[test] +        sealed abstract class List[+A] extends AbstractSeq[A] with LinearSeq[A] with Product with GenericTraversableTemplate[A,List] with LinearSeqOptimized[A,List[A]] with Serializable
[test]  
[test]        )
[test]        args = List(
[test] @@ -146,7 +146,7 @@ TypeRef(
[test]        args = List(
[test]          TypeRef(
[test]            TypeSymbol(
[test] -            sealed abstract class List[+A] extends AbstractSeq[A] with LinearSeq[A] with Product with GenericTraversableTemplate[A,List] with LinearSeqOptimized[A,List[A]] with FilteredTraversableInternal[A,List[A]] with Serializable
[test] +            sealed abstract class List[+A] extends AbstractSeq[A] with LinearSeq[A] with Product with GenericTraversableTemplate[A,List] with LinearSeqOptimized[A,List[A]] with Serializable
[test]  
[test]            )
[test]            args = List(
[test] @@ -179,7 +179,7 @@ PolyType(
[test]            args = List(
[test]              TypeRef(
[test]                TypeSymbol(
[test] -                sealed abstract class List[+A] extends AbstractSeq[A] with LinearSeq[A] with Product with GenericTraversableTemplate[A,List] with LinearSeqOptimized[A,List[A]] with FilteredTraversableInternal[A,List[A]] with Serializable
[test] +                sealed abstract class List[+A] extends AbstractSeq[A] with LinearSeq[A] with Product with GenericTraversableTemplate[A,List] with LinearSeqOptimized[A,List[A]] with Serializable
[test]  
[test]                )
[test]                args = List(TypeParamTypeRef(TypeParam(T <: AnyVal)))
[test] @@ -202,7 +202,7 @@ PolyType(
[test]      params = List(TermSymbol(x: T), TermSymbol(y: List[U]))
[test]      resultType = TypeRef(
[test]        TypeSymbol(
[test] -        sealed abstract class List[+A] extends AbstractSeq[A] with LinearSeq[A] with Product with GenericTraversableTemplate[A,List] with LinearSeqOptimized[A,List[A]] with FilteredTraversableInternal[A,List[A]] with Serializable
[test] +        sealed abstract class List[+A] extends AbstractSeq[A] with LinearSeq[A] with Product with GenericTraversableTemplate[A,List] with LinearSeqOptimized[A,List[A]] with Serializable
[test]  
[test]        )
[test]        args = List(TypeParamTypeRef(TypeParam(U >: T)))
[test] 
...
```

Could be addressed by updating the checkfile `sbt "partest --update-check run/repl-colon-type.scala`

But I question why you're introducing `FilteredTraversableInternal` as part of this change? Is it still relevant in the 2.13.x branch where we don't have binary compatiblity constraints? I'd suggest just adding the new implementations to `List` directly.

**Comment by retronym** [comment]
Minimizes to:

```
object TreeSetTest {
  def test(subject: scala.collection.immutable.TreeSet[Int]) = subject.zip(subject)
}
```

```
⚡ qscalac  sandbox/test.scala
sandbox/test.scala:2: error: ambiguous implicit values:
 both method newCanBuildFrom in object SortedSet of type [A](implicit ord: Ordering[A])scala.collection.generic.CanBuildFrom[scala.collection.immutable.SortedSet.Coll,A,scala.collection.immutable.SortedSet[A]]
 and method newCanBuildFrom in class SortedSetFactory of type [A](implicit ord: Ordering[A])scala.collection.generic.CanBuildFrom[scala.collection.immutable.TreeSet.Coll,A,scala.collection.immutable.TreeSet[A]]
 match expected type scala.collection.generic.CanBuildFrom[scala.collection.immutable.TreeSet[Int],(Int, Int),That]
  def test(subject: scala.collection.immutable.TreeSet[Int]) = subject.zip(subject)
                                                                          ^
one error found
```

I haven't minimized whatever it is that makes this sometimes pass and sometimes fail.

**Comment by cryptom-dev** [comment]
@retronym okay, I will do similar thing as https://github.com/scala/scala/pull/6372

Thank you for looking at this

**Comment by cryptom-dev** [comment]
performance:

baseline 2.13.x 90ffe626ecd7b888794740f4b4b143d4728d8232
```
[info] # Run complete. Total time: 00:31:25
[info] 
[info] Benchmark                            (size)  Mode  Cnt      Score      Error  Units
[info] ListBenchmark.filter_exc_last             0  avgt   20     39.964 ±   11.377  ns/op
[info] ListBenchmark.filter_exc_last             1  avgt   20    108.877 ±   40.773  ns/op
[info] ListBenchmark.filter_exc_last            10  avgt   20    419.570 ±   79.296  ns/op
[info] ListBenchmark.filter_exc_last           100  avgt   20   2711.160 ± 1578.897  ns/op
[info] ListBenchmark.filter_exc_last          1000  avgt   20  17745.196 ± 3647.482  ns/op
[info] ListBenchmark.filter_exc_mid              0  avgt   20     22.168 ±    9.908  ns/op
[info] ListBenchmark.filter_exc_mid              1  avgt   20     25.270 ±    5.292  ns/op
[info] ListBenchmark.filter_exc_mid             10  avgt   20    139.599 ±    6.421  ns/op
[info] ListBenchmark.filter_exc_mid            100  avgt   20   1321.888 ±   30.671  ns/op
[info] ListBenchmark.filter_exc_mid           1000  avgt   20  12975.399 ±  527.259  ns/op
[info] ListBenchmark.filter_excludeAll           0  avgt   20     15.082 ±    0.868  ns/op
[info] ListBenchmark.filter_excludeAll           1  avgt   20     21.121 ±    1.544  ns/op
[info] ListBenchmark.filter_excludeAll          10  avgt   20     66.097 ±    2.047  ns/op
[info] ListBenchmark.filter_excludeAll         100  avgt   20    530.223 ±    5.490  ns/op
[info] ListBenchmark.filter_excludeAll        1000  avgt   20   5158.415 ±  466.000  ns/op
[info] ListBenchmark.filter_from_mid             0  avgt   20     16.576 ±    1.393  ns/op
[info] ListBenchmark.filter_from_mid             1  avgt   20     30.414 ±    2.772  ns/op
[info] ListBenchmark.filter_from_mid            10  avgt   20    119.492 ±    6.626  ns/op
[info] ListBenchmark.filter_from_mid           100  avgt   20    965.398 ±   22.339  ns/op
[info] ListBenchmark.filter_from_mid          1000  avgt   20  10061.495 ±  776.855  ns/op
[info] ListBenchmark.filter_includeAll           0  avgt   20     14.516 ±    0.651  ns/op
[info] ListBenchmark.filter_includeAll           1  avgt   20     27.210 ±    1.612  ns/op
[info] ListBenchmark.filter_includeAll          10  avgt   20    156.706 ±   28.959  ns/op
[info] ListBenchmark.filter_includeAll         100  avgt   20   1456.659 ±  123.793  ns/op
[info] ListBenchmark.filter_includeAll        1000  avgt   20  12456.409 ± 1112.700  ns/op
[info] ListBenchmark.filter_only_last            0  avgt   20     16.045 ±    0.169  ns/op
[info] ListBenchmark.filter_only_last            1  avgt   20     28.803 ±    0.403  ns/op
[info] ListBenchmark.filter_only_last           10  avgt   20     81.480 ±    0.802  ns/op
[info] ListBenchmark.filter_only_last          100  avgt   20    608.914 ±   13.134  ns/op
[info] ListBenchmark.filter_only_last         1000  avgt   20   6126.541 ±  104.027  ns/op
```

This PR Optimization
```
[info] # Run complete. Total time: 00:31:27
[info] 
[info] Benchmark                            (size)  Mode  Cnt      Score      Error  Units
[info] ListBenchmark.filter_exc_last             0  avgt   20      4.554 ±    1.319  ns/op
[info] ListBenchmark.filter_exc_last             1  avgt   20      6.201 ±    0.228  ns/op
[info] ListBenchmark.filter_exc_last            10  avgt   20    189.247 ±   10.173  ns/op
[info] ListBenchmark.filter_exc_last           100  avgt   20   1847.792 ±   40.684  ns/op
[info] ListBenchmark.filter_exc_last          1000  avgt   20  14454.780 ± 6192.910  ns/op
[info] ListBenchmark.filter_exc_mid              0  avgt   20      4.751 ±    1.502  ns/op
[info] ListBenchmark.filter_exc_mid              1  avgt   20      6.071 ±    0.113  ns/op
[info] ListBenchmark.filter_exc_mid             10  avgt   20     62.440 ±   12.901  ns/op
[info] ListBenchmark.filter_exc_mid            100  avgt   20   1070.520 ±   11.481  ns/op
[info] ListBenchmark.filter_exc_mid           1000  avgt   20   5155.931 ±  212.738  ns/op
[info] ListBenchmark.filter_excludeAll           0  avgt   20      3.860 ±    0.118  ns/op
[info] ListBenchmark.filter_excludeAll           1  avgt   20      6.280 ±    0.714  ns/op
[info] ListBenchmark.filter_excludeAll          10  avgt   20     20.656 ±    1.195  ns/op
[info] ListBenchmark.filter_excludeAll         100  avgt   20    223.719 ±    3.060  ns/op
[info] ListBenchmark.filter_excludeAll        1000  avgt   20   2663.433 ±   74.063  ns/op
[info] ListBenchmark.filter_from_mid             0  avgt   20      3.673 ±    0.114  ns/op
[info] ListBenchmark.filter_from_mid             1  avgt   20      5.027 ±    0.094  ns/op
[info] ListBenchmark.filter_from_mid            10  avgt   20    119.516 ±   15.875  ns/op
[info] ListBenchmark.filter_from_mid           100  avgt   20   1133.278 ±   85.667  ns/op
[info] ListBenchmark.filter_from_mid          1000  avgt   20  10803.648 ±  256.929  ns/op
[info] ListBenchmark.filter_includeAll           0  avgt   20      4.131 ±    0.345  ns/op
[info] ListBenchmark.filter_includeAll           1  avgt   20      5.277 ±    0.771  ns/op
[info] ListBenchmark.filter_includeAll          10  avgt   20     22.902 ±    3.330  ns/op
[info] ListBenchmark.filter_includeAll         100  avgt   20    261.138 ±    4.574  ns/op
[info] ListBenchmark.filter_includeAll        1000  avgt   20   2669.859 ±   99.142  ns/op
[info] ListBenchmark.filter_only_last            0  avgt   20      4.041 ±    0.258  ns/op
[info] ListBenchmark.filter_only_last            1  avgt   20      5.777 ±    1.229  ns/op
[info] ListBenchmark.filter_only_last           10  avgt   20     29.416 ±    1.619  ns/op
[info] ListBenchmark.filter_only_last          100  avgt   20    277.224 ±   27.001  ns/op
[info] ListBenchmark.filter_only_last         1000  avgt   20   2605.717 ±   65.524  ns/op
```

File: src/library/scala/collection/immutable/List.scala [hunk 1]
```diff
@@ -391,6 +391,86 @@ sealed abstract class List[+A] extends AbstractSeq[A]
     }
   }
 
+  override def filter(p: A => Boolean): Self = filterImpl(p, isFlipped = false)
```
Comment (retronym) [inline src/library/scala/collection/immutable/List.scala]: Why not also override `filterNot`?

**Comment by retronym** [comment]
So the core change here is avoiding creating new `::` instances for the unchanged suffix of lists.

This is observable to applications that rely on object identity in some way:

```
⚡ scala -e 'val x = List(1); println(x eq x.filter(_ => true))'
false

⚡ scala-pr 6386 -e 'val x = List(1); println(x eq x.filter(_ => true))'
true
```

List actually has two versions the map method to support both use cases:

```
⚡ scala -e 'val x = List("1"); println(x eq x.mapConserve(identity))'
true

⚡ scala -e 'val x = List("1"); println(x eq x.map(identity))'
false
```

So a more conservative change would be to introduce the new behaviour in a new method `filterConserve`, and rely on callers to opt-in to that when needed.

I'm not sure what's the right approach here. @szeiger Have you thought about this sort of thing in the collections rework?

**Comment by cryptom-dev** [comment]
@retronym you are right! I haven't notice such behavior. I would like hear @szeiger 's idea.
I updated the filterNot method, sorry for missing that

**Comment by rorygraves** [comment]
As these are immutable data structure I think the new behaviour is much more preferable over adding a filterConserve.   
From a technical perspective we generate less garbage and gain from structural sharing (and performance) 

I agree it changes the behaviour, but I think it is in a positive way for the user. 

if(x.filter(...) eq x)). Is a nice behaviour. 

We could add to the scaladoc to define the behaviour.

**Comment by retronym** [comment]
I wonder if someone could share the historical reason for the split between `map` and `mapConserve`?

It has a different signature that means it
 - can only be used for endofunctions `e.g. mapping `String => String`)
 - where the element type is `<: AnyRef`
 - and no custom `CanBuildFrom` is allowed

But AFAICT you could safely and soundly make the regular `List.map` implenent the same structural sharing trick by casting the `A` and `B` to `AnyRef` and making the `eq` comparison, and only doing any of this when we're dealing with the default `CanBuildFrom`. 

`filterConserve` doesn't require a different signature (once we're happy to `.asInstanceOf[AnyRef]` in the implementation), so the only argument to keep it separate from `filter` would be backwards compatibility. 2.13 does seem like the right time to make this change.

File: src/library/scala/collection/immutable/List.scala (hunk 1 above)
Comment (mkeskells) [reply src/library/scala/collection/immutable/List.scala]: I think it is on line 396

**Comment by retronym** [comment]
tl;dr; annotate the type of `filter` and `filterNot` as `List[A]` rather than `Self`.

---------------

The strange test failure is related to a change in the order the the decls of `ValueClassDefinitions` are reconstituted after pickling into the `ScalaSignature` and then unpicking back into a `ClassInfoType`. This process does not always preserve the source file order, although this will be fixed as of #6285.

The subtle change is triggered by the fact that this PR declares the return type as `List.this.Self`, rather than just `List[A]`.

In some cases (TODO: is this a bug?), that type isn't dealiased in the inferred type of callers of `filter`.

```
scala> trait B { val test = test2.filterNot(???); val test2: List[String] = Nil }
defined trait B

scala> typeOf[B].decls.map(_.info)
res20: Iterable[reflect.runtime.universe.Type] = List(()Unit, => B.this.test2.Self, => List[String])
```

Or in the case that's causing the code generation to change:

```
diff --git a/Definitions.class.scalap b/Definitions.class.scalap
index eca2645..1c8ec0f 100644
--- a/Definitions.class.scalap
+++ b/Definitions.class.scalap
@@ -563,9 +563,9 @@
     val FloatTpe: scala.reflect.internal.Types.Type = { /* compiled code */ }
     val DoubleTpe: scala.reflect.internal.Types.Type = { /* compiled code */ }
     val BooleanTpe: scala.reflect.internal.Types.Type = { /* compiled code */ }
-    val ScalaNumericValueClasses: scala.collection.immutable.List[scala.reflect.internal.Symbols.ClassSymbol] = { /* compiled code */ }
-    val ScalaValueClassesNoUnit: scala.collection.immutable.List[scala.reflect.internal.Symbols.ClassSymbol] = { /* compiled code */ }
+    val ScalaNumericValueClasses: scala.collection.TraversableLike.Self = { /* compiled code */ }
     val ScalaValueClasses: scala.List[scala.reflect.internal.Symbols.ClassSymbol] = { /* compiled code */ }
+    val ScalaValueClassesNoUnit: scala.collection.TraversableLike.Self = { /* compiled code */ }
     def ScalaPrimitiveValueClasses: scala.List[scala.reflect.internal.Symbols.ClassSymbol] = { /* compiled code */ }
     def underlyingOfValueClass(clazz: scala.reflect.internal.Symbols.Symbol): scala.reflect.internal.Types.Type = { /* compiled code */ }
     private[Definitions] def catastrophicFailure(): scala.Nothing = { /* compiled code */ }
```

**Comment by retronym** [comment]
I took the liberty of `push -f`-ing this change to your `2.13.x_filter_opt` branch.

**Comment by cryptom-dev** [comment]
sure, thank you @retronym

**Review by retronym — APPROVED** [review]
